contrib/99designs/gqlgen/tracer.go

```diff
@@ -0,0 +1,140 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2019 Datadog, Inc.
+
+// Package gqlgen contains an implementation of a gqlgen tracer, and functions to construct and configure the tracer.
+// The tracer can be passed to the gqlgen handler (see package github.com/99designs/gqlgen/handler)
+package gqlgen
+
+import (
+	"context"
+	"fmt"
+	"math"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+
+	"github.com/99designs/gqlgen/graphql"
+)
+
+const (
+	tagResolverObject = "resolver.object"
+	tagResolverField  = "resolver.field"
+)
+
+const defaultResourceName = "gqlgen"
+
+type gqlTracer struct {
+	cfg config
+}
+
+// New creates an a graphql.Tracer instance that can be passed to a gqlgen handler.
+// Options can be passed in for further configuration.
+func New(opts ...Option) graphql.Tracer {
+	var t gqlTracer
+	defaults(&t.cfg)
+	for _, o := range opts {
+		o(&t.cfg)
+	}
+	return &t
+}
+
+// gqlTracer implements the graphql.Tracer interface.
+func (t *gqlTracer) StartOperationParsing(ctx context.Context) context.Context {
+	// not implemented
+	return ctx
+}
+
+// gqlTracer implements the graphql.Tracer interface.
+func (t *gqlTracer) EndOperationParsing(ctx context.Context) {
+	// not implemented
+}
+
+// gqlTracer implements the graphql.Tracer interface.
+func (t *gqlTracer) StartOperationValidation(ctx context.Context) context.Context {
+	// not implemented
+	return ctx
+}
+
+// gqlTracer implements the graphql.Tracer interface.
+func (t *gqlTracer) EndOperationValidation(ctx context.Context) {
+	// not implemented
+}
+
+func (t *gqlTracer) StartOperationExecution(ctx context.Context) context.Context {
+	rctx := graphql.GetRequestContext(ctx)
+	name := defaultResourceName
+	if rctx != nil && rctx.OperationName != "" {
+		name = rctx.OperationName
+	}
+	opts := []ddtrace.StartSpanOption{
+		tracer.SpanType(ext.SpanTypeGraphql),
+		tracer.ResourceName(name),
+		tracer.ServiceName(t.cfg.serviceName),
+	}
+	if !math.IsNaN(t.cfg.analyticsRate) {
+		opts = append(opts, tracer.Tag(ext.EventSampleRate, t.cfg.analyticsRate))
+	}
+	if s, ok := tracer.SpanFromContext(ctx); ok {
+		opts = append(opts, tracer.ChildOf(s.Context()))
+	}
+	_, ctx = tracer.StartSpanFromContext(ctx, name, opts...)
```

Comment:
Unfortunately root spans at Datadog need to have constant operation names, because the application selects the most used one and displays the services and resources for that one alone. If the operation name changes across requests, some traces (and/or resources/services) will not show up in the UI which may be annoying and confusing to users. This is something we are working on and have solutions for, but we are not quite there yet.

If `*ResourceContext.OperationName` is known, let's use it as the resource, and let's have the name be constant, perhaps `"gqlgen.operation"` (or something like that)...